cpp/cmake/thirdparty/get_spdlog.cmake

```diff
@@ -1,27 +0,0 @@
-# =============================================================================
```

Comment:
If I'm asking a question that's already been answered somewhere else, please just give me links and tell me to go read them.

Why is the `fmt` dependency here (`fmt` host dep in `libcudf` conda recipe, `get_fmt.cmake` invoked from `cpp/CMakeLists.txt`) not also being removed? I don't see any explicit uses of `fmt` across this repo... just the ones that'll be generated by `rapids-logger`. I looked through https://github.com/rapidsai/rapids-logger and it doesn't seem to be managing its own `fmt` dependency but.... shouldn't it?

Reply:
Let me know if [this comment in rmm](https://github.com/rapidsai/rmm/blob/branch-25.02/cmake/thirdparty/get_spdlog.cmake#L20) clears things up.

Reply:
It doesn't, sorry. I'd seen that same comment in `rapids-logger` ([code link](https://github.com/rapidsai/rapids-logger/blob/14bb233d2420f7187a690f0bb528ec0420c70d48/get_spdlog.cmake#L20)).

* it seemed to be answering: *how do we make `spdlog` play nicely with a conda-installed `fmt`?*
* but I'm asking: *why even conda install `fmt`, if we could just instead CPM download it in `rapids-logger`?*

Is the answer "because `fmt` is probably going to end up in the build environment anyway as a dependency of something else, and this is the only/easiest way we can ensure the version we want is found/used"?

Reply:
Right OK, that is a bit more complicated. You're on the right track. The problem is actually that spdlog could wind up in the build environment regardless of what we specify if some other package in our build tree needs it. If we have a version of spdlog that is incompatible with cudf, then everything works fine because rapids-cmake will trigger downloading a copy of the right version of spdlog and all is well with the world. If the installed version happens to be compatible with our constraints, though, then the spdlog that we find will be one that causes us to link to libfmt. If we do not include fmt in our dependency list, then we won't have a run export on it, which means that the environment in which cudf is installed may not have fmt (or could have the wrong version of fmt), and then we'd get runtime errors.

Reply:
Thanks for that!

I think I'm just 1 question/answer away from fully getting this.

> _...If we have a version of spdlog that is incompatible with cudf, then everything works fine because rapids-cmake will trigger downloading a copy of the right version of spdlog and all is well with the world._

Then why not just `CPM_DOWNLOAD_spdlog=ON CPM_DOWNLOAD_fmt=ON` in `rapids-logger` to force it to download?

If I'm understanding correctly:

* would guarantee we don't fall into the "got `spdlog` from the conda environment, and therefore linked to `libfmt`" situation
* shouldn't result in any clobbering issues because `cudf_logger_impl` is a `PRIVATE`, header-only dependency
* shouldn't result in any runtime issues because we're not letting any `spdlog::` or `fmt::` symbols into `libcudf`'s public API

Reply:
Good question! This is what I thought you might ask.

You are right, that would work for cudf. However, that would not work for rmm (or raft, or more generally and library that is itself header-only and exports a logger). The problem is that there isn't a good way right now to propagate the `CPM_DOWNLOAD_spdlog` to consumers. Concretely, the rmm logger is not built by rmm, it's a source file that consumers of rmm (such as cudf) build when they do `target_link_libraries(<tgt> PRIVATE rmm::rmm_logger_impl)`. That means that whoever is building `tgt` needs to set this option, not rmm itself. For projects like cudf that have their own logger in addition to rmm's, we'd be setting it anyway, but for projects like cugraph that don't we would have to propagate this setting everywhere to ensure that we get the desired linkage.

One naive solution would be to make it possible to hard code `CPM_DOWNLOAD_spdlog` into rmm's own config file. However, that wouldn't work because we don't include CPM commands in the config files that we install, only the ones in the build directory (this is one of the differences between a `BUILD_EXPORT_SET` and an `INSTALL_EXPORT_SET`). The reason for that is that we typically don't want to write CMake files that end-users of our libraries (including people building their own libraries against our packages) to download something; they should be getting the libraries on their own. We could modify this behavior, but then we would _force_ every consumer of rmm to download spdlog whenever they build their own library, which might not be what they want if they already have a copy on their system. Recall that the main reason we care as much as we do is because we are trying to build packages that are independent of the version of spdlog/fmt on our consumers system, but for end-users of our packages that doesn't matter.

The better solution is to have our installed CMake config files also support configuration of whether or not to force downloading of spdlog. If someone is building a leaf node in our build tree like cugraph and they build the entire tree from source using CPM, there is of course no way to configure on a per-subproject (e.g. raft, rmm) basis whether or not to download spdlog; it's all or nothing for the entire build. In the more typical case where cugraph is being built against pre-built rmm/raft libraries, though, cugraph's build will be able to control whether the copy of the rmm logger it builds during its own build is going to download spdlog or not by setting an option that rmm's rmm-config.cmake file will handle appropriate to decide where to get spdlog from. This is precisely what https://github.com/rapidsai/rmm/pull/1740/ is for.

Reply:
Thank you for taking the time to write that out! I feel like I totally understand the current state and what's coming next now. And why `fmt` needs to stay in cudf's conda dependencies *for now*.